lib/puppet/functions/defined.rb

```diff
@@ -0,0 +1,130 @@
+# (including undef...as opposed to never having been assigned anything). Returns true
+# or false. Accepts class names, type names, resource references, and variable
+# reference strings of the form '$name'.  When more than one argument is
+# supplied, defined() returns true if any are defined.
+#
+# The `defined` function checks both native and defined types, including types
+# provided as plugins via modules. Types and classes are both checked using their names:
+#
+#     defined("file")
+#     defined("customtype")
+#     defined("foo")
+#     defined("foo::bar")
+#     defined('$name')
+#
+# Resource declarations are checked using resource references, e.g.
+# `defined( File['/tmp/myfile'] )`, or `defined( Class[myclass] )`.
+# Checking whether a given resource
+# has been declared is, unfortunately, dependent on the evaluation order of
+# the configuration, and the following code will not work:
+#
+#     if defined(File['/tmp/foo']) {
+#         notify { "This configuration includes the /tmp/foo file.":}
+#     }
+#     file { "/tmp/foo":
+#         ensure => present,
+#     }
+#
+# However, this order requirement refers to evaluation order only, and ordering of
+# resources in the configuration graph (e.g. with `before` or `require`) does not
+# affect the behavior of `defined`.
+#
+# You may also search using types:
+#
+#     defined(Resource['file','/some/file'])
+#     defined(File['/some/file'])
+#     defined(Class['foo'])
+#
+# The `defined` function does not answer if data types (e.g. `Integer`) are defined.
+#
+# The rules for asking for undef, empty strings, and the main class are different from 3.x
+# (non future parser) and 4.x (with future parser or in Puppet 4.0.0 and later):
+#
+#     defined('')     # 3.x => true,  4.x => false
+#     defined(undef)  # 3.x => true,  4.x => error
+#     defined('main') # 3.x => false, 4.x => true
+#
+# With the future parser, it is also possible to ask specifically if a name is
+# a resource type (built in or defined), or a class, by giving its type:
+#
+#     defined(Type[Class['foo']])
+#     defined(Type[Resource['foo']])
+#
+# Which is different from asking:
+# 
+#     defined('foo')
+#
+# Since the later returns true if 'foo' is either a class, a built-in resource type, or a user defined
+# resource type, and a specific request like `Type[Class['foo']]` only returns true if `'foo'` is a class.
+#
+# @since 2.7.0
+# @since 3.6.0 variable reference and future parser types")
+# @since 3.8.1 type specific requests with future parser
+#
+Puppet::Functions.create_function(:'defined', Puppet::Functions::InternalFunction) do
+
+  ARG_TYPE = 'Variant[String,Type[CatalogEntry], Type[Type[CatalogEntry]]]'
+
+  dispatch :is_defined do
+    scope_param
+    optional_repeated_param ARG_TYPE, 'additional_args'
+  end
+
+  def is_defined(scope, *vals)
+    vals.any? do |val|
+      case val
+      when String
+        if m = /^\$(.+)$/.match(val)
+          scope.exist?(m[1])
+        else
```

Comment:
it also avoids one method call (match), the operators are faster than method calls.